### ISSUE TITLE
How to use the EDF reader as an input

### ISSUE THREAD

**Issue by encoded-evolution** [neuromore/studio#193]
Hi, I have made a very simple classifier using an EDF reader as an input, a channel splitter, and I have a Viewer module attached to one of the channels.  No errors are showing.  And the channel splitter correctly shows the channels in the EDF file.  So far, so good.

Yet, I cannot switch from the Test Device to Use Active Device. When I try to do so, it automatically turns on all Test Device channels and uses the auto-generated signals instead.  I cannot get it to stick to "Use Active Device".  When I shut off Test Device in Settings, it shuts off, and Active Device is the default, but then nothing shows in the graphs at all.

Even when I hit GO to start session, same thing occurs.  Apparently reading from an EDF is not a "device".  How do I use stored data from an EDF for analysis?  How do I troubleshoot this?

note: For the EDF reader node, I am using the default sample rate of 0.0, and it automatically finds 256.0

![image](https://user-images.githubusercontent.com/1551406/193421611-a38eaeb4-e17a-4ee2-b789-d35c143f3d3a.png)

Another view with more info.  The file is obviously being read.  Yet nothing shows below

![image](https://user-images.githubusercontent.com/1551406/193422151-388a09f5-f79a-42c2-a3f7-fc60424d6e79.png)

**Comment by stellarpower** [comment]
I don't think files, OSC, etc. are currently supported as first-class devices. The way I have inspected data like this is just to use the signal view node; but not sure why this is not working. Sample rate of 0 sounds interesting...

**Comment by encoded-evolution** [comment]
> I don't think files, OSC, etc. are currently supported as first-class devices. The way I have inspected data like this is just to use the signal view node; but not sure why this is not working. Sample rate of 0 sounds interesting...

Yes, I don't understand why it wouldn't work either.  I'd be happy to try other ideas.

**Comment by JuliusCosmoRomeo** [comment]
Hey @encoded-evolution, excuse the late reply. Good catch, it seems that the EDF Reader node is not working at the moment. We've fixed the EDF Writer node last year to enable recording of files but in that context didn't touch the Reader. 
We'll add it to our backlog and will provide a fix for this once we have capacity and resources. Thanks for your understanding!
Julius

**Comment by stellarpower** [comment]
Is this still expected to be reproducible, I presume? 

On 1.7.1, if I add a file reader node, I just get a constant zero trace in the signal view. It's a file that I originally output using the writer node aswell. I've checked the sample rate I've provided against what EDFBrowser shows me in the metadata. Don't get anything in the logs:
```
2023-07-08 20:56:03.000: Loading channels in EDF(+) / BDF(+) format...
2023-07-08 20:56:05.000: Done.
2023-07-08 20:56:05.000: Read 1 samples in 1 channels.
```

On [another file (first zip)](https://www.teuniz.net/edf_bdf_testfiles/) the process crashes.



I'll see if I can break out a debugger later.

**Comment by stellarpower** [comment]
Working on a fix now. If I don't make it, cause it's pretty slow to debug the handrolled STL types that my debugger can't print for me, I think the issue is starting on [this line]( https://github.com/neuromore/studio/blob/1.7.3/src/Engine/DSP/ChannelFileReader.cpp#L308  ):
```C++
channel->ForceUpdateSampleCounters();
```

The indexing is one dimension out - mNumSamples comes out as 1, but the container below has the value I expect. This call isn't made when we read in CSV.

**Comment by stellarpower** [comment]
Have "v2 channels" been implemented? I am guessing [this comment]( https://github.com/neuromore/studio/blob/1.7.3/src/Engine/DSP/ChannelFileReader.cpp#L300  ) probably indicates our culprit.
